### PR TITLE
Subject Group Viewer - Mar 2020 UI update

### PR DESCRIPTION
**Pull request by shaunanoordin**
## PR Overview

Follows #2036 

We've got additional UI/UX feedback from Chris in regards to the UI changes in 2036. The mint-green focus border is now _too_ obvious for a focus indicator; it's too easily confused for the "cell is selected" indicator. From this feedback, two things need to be done:

1. I need to update the visuals for the focus indicator again. The focus indicator needs to be... fainter? More distinct?
2. I need to revise the review process so I can get better & faster feedback from design (Becky) and user (Chris).

This will probably mean I need to push updates to the preview URL.

### Dev Notes

At the moment, getting an "external glow" (i.e. any visual effect outside the cell's shape) is an absolute pain due to z-index layering. (TL;DR: SVG elements don't support the zIndex CSS or anything similar. You have to rearrange the order of cells to put something "on top".)

One positive thing I can note is that even faint visual indicators (for keyboard focus) can be noticeable IF there's motion. **Still screenshots won't cut it.** Gonna figure out how to capture them GIFs, brb.

### Testing

Local testing URLs:
- Survos Survey: https://local.zooniverse.org:8080/?env=staging&project=1900&workflow=3412
- Galaxy Zoo X: https://local.zooniverse.org:8080/?env=production&project=14993&workflow=17619

Online preview:
- TBC

### Status

WIP. I'm testing a few things to see what I can do, and then I'll push them to preview and tag Becky, Chris, and dev review.

### CODE REVIEW

**Comment by shaunanoordin** [comment]
## PR Update

New design:
- Keyboard focus is now a _transparent white dashed_ outline. 
- The effect is very subtle on a 5x5 but _still noticeable_ when there's movement, which is a key factor in keyboard tab navigation.

Preview videos:

- https://drive.google.com/file/d/1NwfZf_BWvjn6DDFsiGlTeSxwRwTZ3iDK/view?usp=sharing - 5x5 black and white images
- https://drive.google.com/file/d/1eOK7AbBuYR2_w3whCNZfDPGRiOrmFLF0/view?usp=sharing - 2x2 colour images.
- 19MB and 14MB videos. Requires Zooniverse account to view.

### Status

I've been informed that this needs to get rolling very quickly, so I'm putting this straight up for review. Tagging @beckyrother for design thoughts, @chrislintott and @camallen for project-related notes.

**Comment by chrislintott** [comment]
@shaunanoordin I think that looks good. Thanks.

**Comment by shaunanoordin** [comment]
Thanks Chris! 👍 We'll get final code review and these changes will be merged.

Note to self: in the future share these videos on the _public share folder;_ there's nothing so sensitive about demonstrative videos that requires a Zooniverse login.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SubjectGroupViewer/SubjectGroupViewerContainer.js [hunk 1]
```diff
@@ -23,8 +23,7 @@ const DEFAULT_CELL_STYLE = {
   overlay: 'rgba(41,103,255,0.3)',
   selectedStroke: '#2967FF',
   selectedStrokeWidth: '8',
-  focusStroke: '#06FE76',
-  focusStrokeWidth: '4',
+  focusOutline: '2px dashed rgba(255, 255, 255, 0.5)',
```
Comment (camallen) [inline packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SubjectGroupViewer/SubjectGroupViewerContainer.js:26]: not blocking: the `focusOutline` (dashed lines) is easy to see on the smaller grids (larger images) and harder to see on the larger grids (smaller images). 

I'm wondering if we can modify the `focusOutline` between a clamping range of say 2..4(or6) based on the number total number of images in a grid via JS? 

Perhaps even just two types:
- large grid `focusOutlineLargeGrid: '4px dashed rgba(255, 255, 255, 0.5)',` (or 6px) 
- small grid `focusOutlineNormalGrid: '2px dashed rgba(255, 255, 255, 0.5)',`

**Review by camallen — APPROVED** [review]
LGTM - nothing blocking, wondering if we can make the larger grids have a more obvious focus outline by a different CSS property? 

Let's get this out for testing in the wild :shipit:

**Comment by eatyourgreens** [comment]
I think `:focus-visible` is in Chrome, Firefox and Edge now so might be worth using, with graceful degradation in Safari.
[Keyboard-only focus styles](https://css-tricks.com/keyboard-only-focus-styles/).

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SubjectGroupViewer/SubjectGroupViewerContainer.js (hunk 1 above)
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SubjectGroupViewer/SubjectGroupViewerContainer.js:26]: That might be possible with a calculated value in CSS too.

Comment (srallen) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SubjectGroupViewer/SubjectGroupViewerContainer.js:26]: There's some funky things you can do [with calc and css variables](https://www.kizu.ru/conditions-for-css-variables/) to get conditional outcomes... you can manage conditionals by way of media queries if one applies here too. In general, we're not leveraging newer CSS spec enough. [And maybe for the future if it actually gets proposed](https://tabatkins.github.io/specs/css-when-else/). Might not be worth it for readability and understandability though, so interpolating some JS into the CSS might be best.

**Review by eatyourgreens — APPROVED** [review]

**Comment by shaunanoordin** [comment]
Thanks for the review, @camallen and @eatyourgreens !

Jim: I experimented with `:focus-visible` and found that _not_ specifying `:focus` actually triggers a _default focus style_ for mouse interaction on Chrome. (Solid white border.) I didn't realise this before! (Must be Chrome recognises an SVG element with tabIndex=0 should have a visual focus indicator?) However, I'm going to keep this as a future improvement to explore, as I'm not satisfied with the inconsistent visuals I'm getting between keyboard vs mouse, on Chrome vs Safari vs Firefox. (I assume Edge is the same as Chrome)

Cam: good news! The `focusOutline` can be configured _per workflow,_ so I can make a 5x5 grid have `4px dashed rgb(255,255,255,0.1)` to compensate for the smaller cells, or `8px dotted magenta` if I'm feeling fabulous. I'm keeping 2px as the default since it looks best on the test projects I've seen so far. 👌 This default can be adjusted in the future, but for now we have a solution on a per-project basis.

Note, config is:
```
myMorkflow.configuration.subject_viewer_config.cell_style = {
  stroke: '#fff',
  strokeWidth: '1',
  overlay: 'rgba(41,103,255,0.3)',
  selectedStroke: '#2967FF',
  selectedStrokeWidth: '8',
  focusOutline: '2px dashed rgba(255, 255, 255, 0.5)',  // This needs to be tweaked according to number of cells, and width/height of cells
  background: '#000'
}
// ⚠️ ALL items in cell_style must be spelled out. Idea for future-proofing would be to use an Object.assign() to instead merge values between the default cell_style and the user-specified one.
```

I'm going to go ahead and merge this so the changes will be ready for next deployment and/or the next user preview session. Thanks all!

**Comment by shaunanoordin** [comment]
Thanks for the review, @camallen and @eatyourgreens !

Jim: I experimented with `:focus-visible` and found that _not_ specifying `:focus` actually triggers a _default focus style_ for mouse interaction on Chrome. (Solid white border.) I didn't realise this before! (Must be Chrome recognises an SVG element with tabIndex=0 should have a visual focus indicator?) However, I'm going to keep this as a future improvement to explore, as I'm not satisfied with the inconsistent visuals I'm getting between keyboard vs mouse, on Chrome vs Safari vs Firefox. (I assume Edge is the same as Chrome)

Cam: good news! The `focusOutline` can be configured _per workflow,_ so I can make a 5x5 grid have `4px dashed rgb(255,255,255,0.1)` to compensate for the smaller cells, or `8px dotted magenta` if I'm feeling fabulous. I'm keeping 2px as the default since it looks best on the test projects I've seen so far. 👌 This default can be adjusted in the future, but for now we have a solution on a per-project basis.

Note, config is:
```
myMorkflow.configuration.subject_viewer_config.cell_style = {
  stroke: '#fff',
  strokeWidth: '1',
  overlay: 'rgba(41,103,255,0.3)',
  selectedStroke: '#2967FF',
  selectedStrokeWidth: '8',
  focusOutline: '2px dashed rgba(255, 255, 255, 0.5)',  // This needs to be tweaked according to number of cells, and width/height of cells
  background: '#000'
}
// ⚠️ ALL items in cell_style must be spelled out. Idea for future-proofing would be to use an Object.assign() to instead merge values between the default cell_style and the user-specified one.
```

I'm going to go ahead and merge this so the changes will be ready for next deployment and/or the next user preview session. Thanks all!

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SubjectGroupViewer/SubjectGroupViewerContainer.js (hunk 1 above)
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SubjectGroupViewer/SubjectGroupViewerContainer.js:26]: Lea Verou's been publishing a lot of good stuff over the past year or so. I think `min()`, `max()` and `clamp()` are in all the major browsers now. This is one of her tutorials with a hybrid JS/CSS approach to conditional styling: https://lea.verou.me/2020/06/hybrid-positioning-with-css-variables-and-max/

**Comment by beckyrother** [comment]
I think this works! Sorry I missed this on Friday.

**Pull request merged**